unidist/core/backends/mpi/core/controller/garbage_collector.py

```diff
@@ -105,15 +108,18 @@ def regular_cleanup(self):
                 (self._cleanup_counter % self._cleanup_threshold) == 0,
             )
         )
+        # clear completed non blocking operations
+        async_operations = AsyncOperations.get_instance()
+        async_operations.check()
         if len(self._cleanup_list) > self._cleanup_list_threshold:
             if self._cleanup_counter % self._cleanup_threshold == 0:
-
                 timestamp_snapshot = time.perf_counter()
                 if (timestamp_snapshot - self._timestamp) > self._time_threshold:
                     logger.debug("Cleanup counter {}".format(self._cleanup_counter))
 
                     mpi_state = communication.MPIState.get_instance()
                     # Compare submitted and executed tasks
+                    # a blocking send is used because completion is necessary for processing to continue
```

Comment:
```suggestion
                    # We use a blocking send here because we have to wait for
                    # completion of the communication, which is necessary for the pipeline to continue.
```